Port _WKWebExtensionRegisteredScriptsSQLiteStore to C++

ItsJamie9494 opened the pull request:
#### 4e060dd91f139db738f7608ea37619d5c87bc17e
<pre>
Port _WKWebExtensionRegisteredScriptsSQLiteStore to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=287751">https://bugs.webkit.org/show_bug.cgi?id=287751</a>

Reviewed by Timothy Hatcher.

Following previous work with C++ SQLite databases, this ports
the Registered Scripts SQLite store to C++ and removes the previous Cocoa-only implementation.

Tests: Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-shm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-wal
       Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm

* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp: Added.
(WebKit::rowFilterStringFromRowKeys):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::WebExtensionRegisteredScriptsSQLiteStore):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::updateScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::deleteScriptsWithIDs):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::addScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getScriptsWithErrorMessage):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::insertScript):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::isDatabaseEmpty):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::databaseURL):
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStoreCocoa.mm: Added.
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateData):
* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStatement.cpp:
(WebKit::WebExtensionSQLiteStatement::fetch):
(WebKit::WebExtensionSQLiteStatement::fetchWithEnumerationCallback):
(WebKit::WebExtensionSQLiteStatement::reset):
(WebKit::WebExtensionSQLiteStatement::invalidate):
(WebKit::WebExtensionSQLiteStatement::bind):
(WebKit::WebExtensionSQLiteStatement::columnNamesToIndicies):
(WebKit::WebExtensionSQLiteStatement::columnNames):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
(WebKit::toWebAPI):
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.h: Removed.
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.mm: Removed.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::toJSONSerialization):
(WebKit::WebExtensionContext::scriptingRegisterContentScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::loadRegisteredContentScripts):
(WebKit::WebExtensionContext::clearRegisteredContentScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::registeredContentScriptsStore): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::registeredContentScriptsStore):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-shm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-wal: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIScripting, MigrateScriptDataToNewFormat)):

Canonical link: <a href="https://commits.webkit.org/301977@main">https://commits.webkit.org/301977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f7cb331c1900cb06dd99e807df798886c462e71

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38462 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134917 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79203 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47918 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55825 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97166 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65075 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130594 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38321 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114340 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77647 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/37124 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32430 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78276 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54305 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41860 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54816 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110695 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50855 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29287 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19956 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54242 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60418 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53477 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56933 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55235 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -278,45 +345,43 @@
 
-    [registeredContentScriptsStore() deleteScriptsWithIDs:createNSArray(ids).get() completionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, ids, completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
-        if (errorMessage) {
+    registeredContentScriptsStore()->deleteScriptsWithIDs(ids, [this, protectedThis = Ref { *this }, ids = WTFMove(ids), completionHandler = WTFMove(completionHandler)](const String& errorMessage) mutable {
```
Using `WTFMove(ids)` on the same line as `ids` makes me anxious — is order guaranteed?. A copy for the lambda is likely fine here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -213,15 +280,15 @@
 
         InjectedContentVector injectedContents;
-        for (auto& parameters : scripts) {
+        for (auto parameters : scripts) {
```
Why not `auto&` still?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -57,6 +57,73 @@
+    return result;
+}
+
+static Vector<Ref<JSON::Object>> toWebAPI(const Vector<WebExtensionRegisteredScriptParameters>& scripts)
```
`toWebAPI` is not the right name here, it was unfortunate that it was reusing the one for the web process here in the UI process. This is more like `toJSONSerialization`, something to signal this is for storage.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -57,6 +57,73 @@
+static NSString * const mainWorld = @"main";
+static NSString * const isolatedWorld = @"isolated";
+
+static Ref<JSON::Array> toWebAPI(const Vector<String>& items)
```
Ditto. `toJSONSerialization`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -169,8 +236,8 @@
 
-    [registeredContentScriptsStore() addScripts:toWebAPI(scripts) completionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, scripts, injectedContentsMap = WTFMove(injectedContentsMap), completionHandler = WTFMove(completionHandler)](NSString *errorMessage) mutable {
-        if (errorMessage) {
+    registeredContentScriptsStore()->addScripts(toWebAPI(scripts), [this, protectedThis = Ref { *this }, scripts = WTFMove(scripts), injectedContentsMap = WTFMove(injectedContentsMap), completionHandler = WTFMove(completionHandler)](const String& errorMessage) mutable {
```
Ditto on `WTFMove(scripts)` on the same line as `toWebAPI(scripts)`. Maybe pull `toWebAPI(scripts)` (as `toJSONSerialization`) out into a variable before.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
```diff
@@ -113,6 +113,14 @@ id toWebAPI(const std::optional<T>& result, UseNullValue useNull = UseNullValue:
+{
+    String jsonString = object.toJSONString();
+    auto *jsonData = [(NSString *)jsonString dataUsingEncoding:NSUTF8StringEncoding];
+    NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingAllowFragments error:nil];
```
I think an implicit `jsonString` will work for the method call, no need to cast, IIRC.
```suggestion
    auto jsonString = object.toJSONString();
    auto *jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    auto *dictionary = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingAllowFragments error:nil];
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+        return;
+    }
+
+    queue()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, ids = WTFMove(ids), completionHandler = WTFMove(completionHandler)]() mutable {
```
All the queue dispatches need to use `crossThreadCopy` instead of `WTFMove`. See https://commits.webkit.org/290441@main.
```suggestion
    queue()->dispatch([weakThis = ThreadSafeWeakPtr { *this }, ids = crossThreadCopy(ids), completionHandler = WTFMove(completionHandler)]() mutable {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+
+        String deleteDatabaseErrorMessage = protectedThis->deleteDatabaseIfEmpty();
+
+        WorkQueue::protectedMain()->dispatch([deleteDatabaseErrorMessage = WTFMove(deleteDatabaseErrorMessage), errorMessage = WTFMove(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
```
Ditto on `crossThreadCopy`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+        for (const auto& script : persistentScripts)
+            protectedThis->insertScript(script, *(protectedThis->database()), errorMessage);
+
+        WorkQueue::protectedMain()->dispatch([errorMessage = WTFMove(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
```
Ditto on `crossThreadCopy`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+
+        String errorMessage;
+        auto scripts = protectedThis->getScriptsWithErrorMessage(errorMessage);
+        WorkQueue::protectedMain()->dispatch([scripts = WTFMove(scripts), errorMessage = WTFMove(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
```
Ditto on `crossThreadCopy` for `errorMessage`. `scripts` should be fine.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+
+        String errorMessage;
+        if (!protectedThis->openDatabaseIfNecessary(errorMessage, false)) {
+            WorkQueue::protectedMain()->dispatch([errorMessage = WTFMove(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
```
Ditto on `crossThreadCopy`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+
+        String errorMessage;
+        if (!protectedThis->openDatabaseIfNecessary(errorMessage, true)) {
+            WorkQueue::protectedMain()->dispatch([errorMessage = WTFMove(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
```
Ditto on `crossThreadCopy`.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+    Vector<String> escapedAndQuotedKeys;
+    for (String key : keys) {
+        String keyWithSingleQuotesEscaped = makeStringByReplacingAll(key, "'"_s, "''"_s);
+        escapedAndQuotedKeys.append(makeString("'"_s, keyWithSingleQuotesEscaped, "'"_s));
+    }
+    return makeStringByJoining(escapedAndQuotedKeys.span(), ", "_s);
```
```suggestion
    }

    return makeStringByJoining(escapedAndQuotedKeys.span(), ", "_s);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,283 @@
+{
+    Vector<String> escapedAndQuotedKeys;
+    for (String key : keys) {
+        String keyWithSingleQuotesEscaped = makeStringByReplacingAll(key, "'"_s, "''"_s);
```
```suggestion
    for (const auto& key : keys) {
        auto keyWithSingleQuotesEscaped = makeStringByReplacingAll(key, "'"_s, "''"_s);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+    assertIsCurrent(queue());
+
+    auto currentDatabaseSchemaVersion = databaseSchemaVersion();
+    if (currentDatabaseSchemaVersion == 1 && currentSchemaVersion == 2) {
```
Same, I think this needs to migrate any v1.

```suggestion
    if (currentDatabaseSchemaVersion == 1) {
```

xeenon replied:
This still needs changed.

xeenon reviewed (changes requested):
Looks good just migration version check.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+    // Only save persistent scripts to storage
+    Vector<Ref<JSON::Object>> persistentScripts;
+    for (const auto& script : scripts) {
+        if (auto persistent = script->getBoolean(persistAcrossSessionsKey); persistent.value())
```
```suggestion
        if (auto persistent = script->getBoolean(persistAcrossSessionsKey); persistent && persistent.value())
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+        if (RefPtr object = value->asObject())
+            results.append(*object);
+        else
+            RELEASE_LOG_ERROR(Extensions, "Failed to deserialize registered content scripts for extension %s", uniqueIdentifier().utf8().data());
```
Need to null check the output of `parseJSON` and use `RefPtr`.
```suggestion
        if (RefPtr value = JSON::Value::parseJSON(script)) {
            if (RefPtr object = value->asObject())
                results.append(*object);
            else
                RELEASE_LOG_ERROR(Extensions, "Failed to deserialize registered content scripts for extension %s", uniqueIdentifier().utf8().data());
        } else
            RELEASE_LOG_ERROR(Extensions, "Failed to parse JSON for registered content scripts for extension %s", uniqueIdentifier().utf8().data());
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+        migrateData();
+
+        setDatabaseSchemaVersion(currentSchemaVersion);
+        return currentSchemaVersion;
+    }
+    return WebExtensionSQLiteStore::migrateToCurrentSchemaVersionIfNeeded();
```
```suggestion
    }

    return WebExtensionSQLiteStore::migrateToCurrentSchemaVersionIfNeeded();
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+        return;
+    }
+
+    queue().dispatchSync([weakThis = ThreadSafeWeakPtr { *this }, persistentScripts, completionHandler = WTFMove(completionHandler)]() mutable {
```
This should not be `dispatchSync()`.

```suggestion
    queue().dispatch([weakThis = ThreadSafeWeakPtr { *this }, persistentScripts, completionHandler = WTFMove(completionHandler)]() mutable {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
```diff
@@ -113,6 +113,14 @@ id toWebAPI(const std::optional<T>& result, UseNullValue useNull = UseNullValue:
+{
+    auto jsonString = object.toJSONString();
+    RetainPtr jsonData = [jsonString.createNSString() dataUsingEncoding:NSUTF8StringEncoding];
+    NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:jsonData.get() options:NSJSONReadingAllowFragments error:nil];
```
`NSJSONReadingAllowFragments` allows results other than `NSDictionary`. Just pass `0` for dictionary.
```suggestion
    NSDictionary *dictionary = [NSJSONSerialization JSONObjectWithData:jsonData.get() options:0 error:nil];
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+{
+    // Only save persistent scripts to storage
+    Vector<Ref<JSON::Object>> persistentScripts;
+    for (const auto& script : scripts) {
```
```suggestion
    for (Ref script : scripts) {
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,303 @@
+
+        ASSERT(errorMessage.isEmpty());
+
+        for (const auto& script : persistentScripts)
```
```suggestion
        for (Ref script : persistentScripts)
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStoreCocoa.mm:
```diff
@@ -0,0 +1,83 @@
+
+void WebExtensionRegisteredScriptsSQLiteStore::migrateData()
+{
+    static NSSet *allowedClasses = [NSSet setWithObjects:NSString.class, NSNumber.class, NSArray.class, NSMutableDictionary.class, nil];
```
Not new, but `NSMutableDictionary` is wrong here. Should be `NSDictionary`.

Best to match `WebExtensionDeclarativeNetRequestSQLiteStore::migrateData()`.

```suggestion
    static NSSet *allowedClasses = [NSSet setWithObjects:NSDictionary.class, NSNumber.class, NSString.class, NSArray.class, nil];
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp:
```diff
@@ -0,0 +1,306 @@
+    assertIsCurrent(queue());
+
+    auto currentDatabaseSchemaVersion = databaseSchemaVersion();
+    if (currentDatabaseSchemaVersion == 1 && currentSchemaVersion == 2) {
```
```suggestion
    if (currentDatabaseSchemaVersion == 1) {
```

ItsJamie9494 committed:
Port _WKWebExtensionRegisteredScriptsSQLiteStore to C++
https://bugs.webkit.org/show_bug.cgi?id=287751

Reviewed by Timothy Hatcher.

Following previous work with C++ SQLite databases, this ports
the Registered Scripts SQLite store to C++ and removes the previous Cocoa-only implementation.

Tests: Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-shm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-wal
       Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm

* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp: Added.
(WebKit::rowFilterStringFromRowKeys):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::WebExtensionRegisteredScriptsSQLiteStore):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::updateScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::deleteScriptsWithIDs):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::addScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getScripts):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getScriptsWithErrorMessage):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::getKeysAndValuesFromRowIterator):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::insertScript):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::currentDatabaseSchemaVersion):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::createFreshDatabaseSchema):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateToCurrentSchemaVersionIfNeeded):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::resetDatabaseSchema):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::isDatabaseEmpty):
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::databaseURL):
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStoreCocoa.mm: Added.
(WebKit::WebExtensionRegisteredScriptsSQLiteStore::migrateData):
* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStatement.cpp:
(WebKit::WebExtensionSQLiteStatement::fetch):
(WebKit::WebExtensionSQLiteStatement::fetchWithEnumerationCallback):
(WebKit::WebExtensionSQLiteStatement::reset):
(WebKit::WebExtensionSQLiteStatement::invalidate):
(WebKit::WebExtensionSQLiteStatement::bind):
(WebKit::WebExtensionSQLiteStatement::columnNamesToIndicies):
(WebKit::WebExtensionSQLiteStatement::columnNames):
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
(WebKit::toWebAPI):
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.h: Removed.
* Source/WebKit/Shared/Extensions/_WKWebExtensionRegisteredScriptsSQLiteStore.mm: Removed.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::toJSONSerialization):
(WebKit::WebExtensionContext::scriptingRegisterContentScripts):
(WebKit::WebExtensionContext::scriptingUpdateRegisteredScripts):
(WebKit::WebExtensionContext::scriptingUnregisterContentScripts):
(WebKit::WebExtensionContext::loadRegisteredContentScripts):
(WebKit::WebExtensionContext::clearRegisteredContentScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::registeredContentScriptsStore): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::registeredContentScriptsStore):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm:
(WebKit::toWebAPI):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIScripting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-shm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RegisteredContentScripts.db-wal: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIScripting, MigrateScriptDataToNewFormat)):

Canonical link: https://commits.webkit.org/301977@main
